compiler/src/iree/compiler/Codegen/Common/TransformDialectStrategies.cpp

```diff
@@ -0,0 +1,608 @@
+/// func.func.
+// TODO: configure patterns.
+static Value buildVectorizeStrategy(ImplicitLocOpBuilder &b, Value funcH) {
+  funcH = b.create<ApplyPatternsOp>(funcH, /*rankReducing=*/true);
+  return b.create<VectorizeOp>(funcH);
+}
+
+/// Post-bufferization mapping to blocks and threads.
+/// Takes a handle to a func.func and returns an updated handle to a
+/// func.func.
+static Value mapToBlockAndThreads(ImplicitLocOpBuilder &b, Value funcH,
+                                  ArrayRef<int64_t> blockSize) {
+  funcH = b.create<ForeachThreadToWorkgroupOp>(funcH);
+  return b.create<MapNestedForeachThreadToGpuThreadsOp>(funcH, blockSize);
+}
+
+/// Post-bufferization vector distribution with rank-reduction.
+/// Takes a handle to a func.func and returns an updated handle to a
+/// func.func.
+static Value distributeVectors(ImplicitLocOpBuilder &b, Value funcH,
+                               int64_t warpSize = 32) {
+  funcH = b.create<ApplyPatternsOp>(funcH, /*rankReducing=*/true);
+  Value ifH = b.create<MatchOp>(funcH, scf::IfOp::getOperationName());
+  ifH = b.create<VectorToWarpExecuteOnLane0Op>(ifH, warpSize);
+  b.create<VectorWarpDistributionOp>(funcH);
+  return funcH;
+}
+
+/// Various handles produced by reduction splitting.
+struct ReductionSplitResult {
+  /// Handle to the leading elementwise operation, may be null if no such
+  /// operation is present.
+  Value leadingEltwiseH;
+  /// Handle to the fill operation feeding the init of a higher-rank
+  /// more-parallel reduction.
+  Value splitFillH;
+  /// Handle to the higher-rank more-parallel reduction.
+  Value splitLinalgH;
+  /// Handle to the final reduction.
+  Value combinerH;
+  /// Handle to the original fill operation, may be null if the operation was
+  /// not re-matched.
+  Value originalFillH;
+  /// Handle to the trailing fill operation, may be null if the operation was
+  /// not re-matched.
+  Value trailingEltwiseH;
+};
+
+/// Builds transform IR requesting to bubble up the "expand_shape" operation
+/// produced as parent of reduction splitting if necessary for fusion of the
+/// leading elementwise operation.
+static ReductionSplitResult buildExpansionBubbleUp(
+    ImplicitLocOpBuilder &b, Value variantH,
+    SplitReductionOp splitReductionTransformOp, bool hasLeadingEltwise,
+    bool hasTrailingEltwise) {
+  ReductionSplitResult result;
+  if (!hasLeadingEltwise) {
+    result.splitFillH = splitReductionTransformOp.getFillOp();
+    result.splitLinalgH = splitReductionTransformOp.getSplitLinalgOp();
+    result.combinerH = splitReductionTransformOp.getCombiningLinalgOp();
+    return result;
+  }
+
+  auto funcH = b.create<MatchOp>(variantH, func::FuncOp::getOperationName());
+  auto applyPatterns = b.create<ApplyPatternsOp>(funcH, /*rankReducing=*/false);
+  applyPatterns->setAttr(applyPatterns.getBubbleCollapseExpandAttrName(),
+                         b.getUnitAttr());
+  std::tie(result.originalFillH, result.splitFillH) =
```

Comment:
can we use structured bindings here for std::tie?

Reply:
can we?
Not sure how to do this for variables that are already declared.

Reply:
I think structured bindings can only be allowed with `auto`, i.e., `auto [a, b, c] = std::make_tuple(1, 2 3);`. In this case, `std::tie` should be used.

https://stackoverflow.com/questions/46764614/why-structured-bindings-only-work-with-auto